File: JAVA/Percolation/PercolationTests.java
import junit.framework.TestCase;

/**
 * Created by D.Kolomiyets on 2/2/14.
 */

public class PercolationTests extends TestCase {
    public void testThat1DimensionalMatrixPercolatesAfterOneOpen() {
        Percolation target = new Percolation(1);

        target.open(1, 1);

        assertTrue(target.percolates());
    }

    public void testDoesNotPercolate2DimensionalMatrixAfterCreation() {
        Percolation target = new Percolation(2);

        assertFalse(target.percolates());
    }

    public void testIsNotOpenAfterCreation() {
        Percolation target = new Percolation(2);

        assertFalse(target.isOpen(1, 1));
    }

    public void testIsNotFullClosedSite() {
        Percolation target = new Percolation(2);

        assertFalse(target.isFull(1, 1));
    }

    public void testIsNotFull() {
        Percolation target = new Percolation(2);

        target.open(2, 2);

        assertFalse(target.isFull(2, 2));
    }

    public void testIsFull() {
        Percolation target = new Percolation(2);

        target.open(1, 1);

        assertTrue(target.isFull(1, 1));
    }

    public void testIsOpenAfterOpen() {
        Percolation target = new Percolation(2);

        target.open(1, 1);

        assertTrue(target.isOpen(1, 1));
    }

    public void testPercolates2DimensionalMatrix() {
        Percolation target = new Percolation(2);

        target.open(1, 1);
        target.open(2, 1);

        assertTrue(target.percolates());
    }

    public void testDoesNotPercolate2DimensionalMatrix() {
        Percolation target = new Percolation(2);

        target.open(1, 1);
        target.open(2, 2);

        assertFalse(target.percolates());
    }

    public void testPercolates3DimensionalMatrix() {
        Percolation target = new Percolation(3);

        target.open(1, 1);
        target.open(2, 1);
        target.open(3, 1);

        assertTrue(target.percolates());
    }

    public void testDoesNotPercolate3DimensionalMatrix() {
        Percolation target = new Percolation(3);

        target.open(1, 1);
        target.open(2, 2);
        target.open(3, 3);

        assertFalse(target.percolates());
    }

    public void testPercolates4DimensionalMatrix() {
        Percolation target = new Percolation(4);

        target.open(1, 1);
        target.open(2, 1);
        target.open(3, 1);
        target.open(4, 1);

        assertTrue(target.percolates());
    }

    public void testDoesNotPercolate4DimensionalMatrix() {
        Percolation target = new Percolation(4);

        target.open(1, 1);
        target.open(2, 2);
        target.open(3, 3);
        target.open(4, 4);

        assertFalse(target.percolates());
    }

    public void testThatOpenShouldThrowExceptionOnZeroIndexes() {
        boolean thrown = false;
        Percolation target = new Percolation(4);

        try {
            target.open(0, 0);
        } catch (Exception e) {
            thrown = e.getClass() == IndexOutOfBoundsException.class;
        }

        assertTrue(thrown);
    }

    public void testThatOpenShouldThrowExceptionOnGreaterThenSizeIndexSpecified() {
        boolean thrown = false;
        Percolation target = new Percolation(4);

        try {
            target.open(5, 1);
        } catch (Exception e) {
            thrown = e.getClass() == IndexOutOfBoundsException.class;
        }

        assertTrue(thrown);
    }

    public void testThatIsOpenShouldThrowExceptionOnGreaterThenSizeIndexSpecified() {
        boolean thrown = false;
        Percolation target = new Percolation(4);

        try {
            target.isOpen(5, 1);
        } catch (Exception e) {
            thrown = e.getClass() == IndexOutOfBoundsException.class;
        }

        assertTrue(thrown);
    }

    public void testThatIsOpenShouldThrowExceptionOnZeroIndexSpecified() {
        boolean thrown = false;
        Percolation target = new Percolation(4);

        try {
            target.isOpen(0, 0);
        } catch (Exception e) {
            thrown = e.getClass() == IndexOutOfBoundsException.class;
        }

        assertTrue(thrown);
    }

    public void testThatIsFullShouldThrowExceptionOnGreaterThenSizeIndexSpecified() {
        boolean thrown = false;
        Percolation target = new Percolation(4);

        try {
            target.isFull(5, 1);
        } catch (Exception e) {
            thrown = e.getClass() == IndexOutOfBoundsException.class;
        }

        assertTrue(thrown);
    }

    public void testThatIsFullShouldThrowExceptionOnZeroIndexSpecified() {
        boolean thrown = false;
        Percolation target = new Percolation(4);

        try {
            target.isFull(0, 0);
        } catch (Exception e) {
            thrown = e.getClass() == IndexOutOfBoundsException.class;
        }

        assertTrue(thrown);
    }

    public void testIsNotFullEvenIfSiteIsOpenAndSystemPercolates() {
        Percolation target = new Percolation(3);

        target.open(3, 1);

        target.open(1, 3);
        target.open(2, 3);
        target.open(3, 3);

        assertFalse(target.isFull(3, 1));
    }

    public void testIfIllegalArgumentExceptionThrownOnZeroSize() {
        boolean thrown = false;
        try {
            new Percolation(0);
        } catch (Exception e) {
            thrown = e.getClass() == IllegalArgumentException.class;
        }

        assertTrue(thrown);
    }
}

File: CS/Percolation/PercolationTests.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Percolation
{
    [TestClass]
    public class PercolationTests
    {
        [TestMethod]
        public void TestThat1DimensionalMatrixPercolatesAfterOneOpen()
        {
            Percolation target = new Percolation( 1 );

            target.Open( 1, 1 );

            Assert.IsTrue( target.Percolates() );
        }

        [TestMethod]
        public void TestDoesNotPercolate2DimensionalMatrixAfterCreation()
        {
            Percolation target = new Percolation( 2 );

            Assert.IsFalse( target.Percolates() );
        }

        [TestMethod]
        public void TestIsNotOpenAfterCreation()
        {
            Percolation target = new Percolation( 2 );

            Assert.IsFalse( target.IsOpen( 1, 1 ) );
        }

        [TestMethod]
        public void TestIsNotFullClosedSite()
        {
            Percolation target = new Percolation( 2 );

            Assert.IsFalse( target.IsFull( 1, 1 ) );
        }

        [TestMethod]
        public void TestIsNotFull()
        {
            Percolation target = new Percolation( 2 );

            target.Open( 2, 2 );

            Assert.IsFalse( target.IsFull( 2, 2 ) );
        }

        [TestMethod]
        public void TestIsFull()
        {
            Percolation target = new Percolation( 2 );

            target.Open( 1, 1 );

            Assert.IsTrue( target.IsFull( 1, 1 ) );
        }

		[TestMethod]
		[ExpectedException( typeof( ArgumentOutOfRangeException ) )]
		public void TestThatIsFullThrowsExceptionOnInvalidFirstIndex()
		{
			Percolation target = new Percolation( 2 );

			target.IsFull( -1, 1 );
		}

		[TestMethod]
		[ExpectedException( typeof( ArgumentOutOfRangeException ) )]
		public void TestThatIsFullThrowsExceptionOnTooBigSecondIndex()
		{
			Percolation target = new Percolation( 2 );

			target.IsFull( 1, 5 );
		}

		[TestMethod]
		[ExpectedException( typeof( ArgumentOutOfRangeException ) )]
		public void TestThatIsFullThrowsExceptionOnTooLowSecondIndex()
		{
			Percolation target = new Percolation( 2 );

			target.IsFull( 1, -5 );
		}

        [TestMethod]
        public void TestIsOpenAfterOpen()
        {
            Percolation target = new Percolation( 2 );

            target.Open( 1, 1 );

            Assert.IsTrue( target.IsOpen( 1, 1 ) );
        }

        [TestMethod]
        public void TestPercolates2DimensionalMatrix()
        {
            Percolation target = new Percolation( 2 );

            target.Open( 1, 1 );
            target.Open( 2, 1 );

            Assert.IsTrue( target.Percolates() );
        }

        [TestMethod]
        public void TestDoesNotPercolate2DimensionalMatrix()
        {
            Percolation target = new Percolation( 2 );

            target.Open( 1, 1 );
            target.Open( 2, 2 );

            Assert.IsFalse( target.Percolates() );
        }

        [TestMethod]
        public void TestPercolates3DimensionalMatrix()
        {
            Percolation target = new Percolation( 3 );

            target.Open( 1, 1 );
            target.Open( 2, 1 );
            target.Open( 3, 1 );

            Assert.IsTrue( target.Percolates() );
        }

        [TestMethod]
        public void TestDoesNotPercolate3DimensionalMatrix()
        {
            Percolation target = new Percolation( 3 );

            target.Open( 1, 1 );
            target.Open( 2, 2 );
            target.Open( 3, 3 );

            Assert.IsFalse( target.Percolates() );
        }

        [TestMethod]
        public void TestPercolates4DimensionalMatrix()
        {
            Percolation target = new Percolation( 4 );

            target.Open( 1, 1 );
            target.Open( 2, 1 );
            target.Open( 3, 1 );
            target.Open( 4, 1 );

            Assert.IsTrue( target.Percolates() );
        }

        [TestMethod]
        public void TestDoesNotPercolate4DimensionalMatrix()
        {
            Percolation target = new Percolation( 4 );

            target.Open( 1, 1 );
            target.Open( 2, 2 );
            target.Open( 3, 3 );
            target.Open( 4, 4 );

            Assert.IsFalse( target.Percolates() );
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentOutOfRangeException))]
        public void TestThatOpenShouldThrowExceptionOnZeroIndexes()
        {
            Percolation target = new Percolation( 4 );

            target.Open( 0, 0 );
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentOutOfRangeException))]
        public void TestThatOpenShouldThrowExceptionOnGreaterThenSizeIndexSpecified()
        {
            Percolation target = new Percolation( 4 );

            target.Open( 5, 1 );
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentOutOfRangeException))]
        public void TestThatIsOpenShouldThrowExceptionOnGreaterThenSizeIndexSpecified()
        {
            Percolation target = new Percolation( 4 );
            
            target.IsOpen( 5, 1 );
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentOutOfRangeException))]
        public void TestThatIsOpenShouldThrowExceptionOnZeroIndexSpecified()
        {
            Percolation target = new Percolation( 4 );

            target.IsOpen( 0, 0 );
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentOutOfRangeException))]
        public void TestThatIsFullShouldThrowExceptionOnGreaterThenSizeIndexSpecified()
        {
            Percolation target = new Percolation( 4 );

            target.IsFull( 5, 1 );
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentOutOfRangeException))]
        public void TestThatIsFullShouldThrowExceptionOnZeroIndexSpecified()
        {
            Percolation target = new Percolation( 4 );

            target.IsFull( 0, 0 );
        }

        [TestMethod]
        public void TestIsNotFullEvenIfSiteIsOpenAndSystemPercolates()
        {
            Percolation target = new Percolation( 3 );

            target.Open( 3, 1 );

            target.Open( 1, 3 );
            target.Open( 2, 3 );
            target.Open( 3, 3 );

            Assert.IsFalse( target.IsFull( 3, 1 ) );
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentException))]
        public void TestIfIllegalArgumentExceptionThrownOnZeroSize()
        {
            new Percolation( 0 );
        }
    }
}
